src/v/datalake/values_protobuf.cc

```diff
@@ -0,0 +1,333 @@
+        }
+        ret->elements.push_back(std::move(result.value()));
+    };
+    co_return value_outcome{std::move(ret)};
+}
+
+ss::future<value_outcome> convert_map(
+  std::optional<parsed::message::field> field,
+  const pb::FieldDescriptor& field_descriptor,
+  proto_descriptors_stack& stack) {
+    auto ret = std::make_unique<iceberg::map_value>();
+    if (!field.has_value()) {
+        co_return ret;
+    }
+
+    auto parsed_map = std::get<parsed::map>(std::move(field.value()));
+    ret->kvs.reserve(parsed_map.entries.size());
+
+    for (auto& [entry_k, entry_v] : parsed_map.entries) {
+        /**
+         * Convert key, if a parsed key is represented by a monostate we assume
+         * it has a default value
+         */
+        value_outcome key_result = co_await ss::visit(
+          std::move(entry_k),
+          [&field_descriptor, &stack](std::monostate) {
+              // default value
+              return single_field_to_value(
+                std::nullopt,
+                *field_descriptor.message_type()->map_key(),
+                stack);
+          },
+          [&field_descriptor, &stack](auto& value) {
+              return single_field_to_value(
+                std::move(value),
+                *field_descriptor.message_type()->map_key(),
+                stack);
+          });
+        if (key_result.has_error()) {
+            co_return key_result.error();
+        }
+
+        std::optional<value_outcome> opt_value_result = co_await ss::visit(
+          std::move(entry_v),
+          [](std::monostate) {
+              return ssx::now(std::optional<value_outcome>());
+          },
+          [&field_descriptor, &stack](auto& value) {
+              return single_field_to_value(
+                       std::move(value),
+                       *field_descriptor.message_type()->map_value(),
+                       stack)
+                .then([](auto v) {
+                    return std::optional<value_outcome>(std::move(v));
+                });
+          });
+        if (opt_value_result.has_value() && opt_value_result->has_error()) {
+            co_return opt_value_result->error();
+        }
+        std::optional<iceberg::value> value;
+        if (opt_value_result.has_value()) {
+            value = std::move(opt_value_result->value());
+        }
+
+        ret->kvs.push_back(iceberg::kv_value{
+          .key = std::move(key_result.value()), .val = std::move(value)});
+    }
+
+    co_return ret;
+}
+
+ss::future<value_outcome> message_field_to_value(
+  std::optional<parsed::message::field> field,
+  const pb::FieldDescriptor& field_descriptor,
+  proto_descriptors_stack& stack) {
+    if (field_descriptor.is_map()) {
+        return convert_map(std::move(field), field_descriptor, stack);
+    }
+
+    if (field_descriptor.is_repeated()) {
+        if (!field.has_value()) {
+            return ssx::now<value_outcome>(
+              std::make_unique<iceberg::list_value>());
+        }
```

Comment:
Seems inconsistent that this check is inlined here for the list, but for the map case it's in the converter. Is there a reason for this? I think I prefer to inline the map one in this method too.